### PR TITLE
P2474R2 views::repeat

### LINKED ISSUE #5616: [2022-07 LWG Motion 29] P2474R2 views::repeat
[P2474R2](https://wg21.link/p2474r2)

### PR DESCRIPTION
**Pull request by jensmaurer**
Fixes #5616
Fixes cplusplus/papers#1155

### CODE REVIEW

File: source/ranges.tex [hunk 1]
```diff
@@ -3202,6 +3210,428 @@
...
+\indexlibrarymember{operator--}{repeat_view::iterator}%
+\begin{itemdecl}
+constexpr @\exposid{iterator}@& operator--();
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\expects
+If \tcode{Bound} is not \tcode{unreachable_sentinel_t},
+\tcode{\exposid{bound_} > 0}.
```
Comment (JohelEGP) [inline source/ranges.tex:3491]: See <TBD>.
```suggestion
$\exposid{bound_} > 0$.
```

File: source/ranges.tex [hunk 2]
```diff
@@ -3202,6 +3210,428 @@
...
+
+\pnum
+\effects
+Equivalent to:
+\begin{codeblock}
+--current_;
```
Comment (JohelEGP) [inline source/ranges.tex:3497]:
```suggestion
--@\exposid{current_}@;
```

File: source/ranges.tex [hunk 3]
```diff
@@ -3202,6 +3210,428 @@
...
+\pnum
+\expects
+If \tcode{Bound} is not \tcode{unreachable_sentinel_t},
+$\exposid{bound_} - \tcode{n} \ge 0$.
+
+\pnum
+\effects
+Equivalent to:
+\begin{codeblock}
+current_ -= n;
```
Comment (JohelEGP) [inline source/ranges.tex:3553]:
```suggestion
@\exposid{current_}@ -= n;
```

File: source/ranges.tex [hunk 4]
```diff
@@ -3202,6 +3210,428 @@
...
+
+\indexlibrarymember{operator==}{repeat_view::iterator}%
+\begin{itemdecl}
+friend constexpr bool operator==(const @\exposid{iterator}@& x, const @\exposid{iterator}@& y);
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\effects
+Equivalent to \tcode{x.\exposid{current_} == y.\exposid{current_}}.
```
Comment (JohelEGP) [inline source/ranges.tex:3577]: Error in the paper.
```suggestion
Equivalent to \tcode{return x.\exposid{current_} == y.\exposid{current_};}.
```

File: source/ranges.tex [hunk 5]
```diff
@@ -3202,6 +3210,428 @@
...
+
+\indexlibrarymember{operator<=>}{repeat_view::iterator}%
+\begin{itemdecl}
+friend constexpr auto operator<=>(const @\exposid{iterator}@& x, const @\exposid{iterator}@& y);
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\effects
+Equivalent to \tcode{x.\exposid{current_} <=> y.\exposid{current_}}.
```
Comment (JohelEGP) [inline source/ranges.tex:3588]: Error in the paper.
```suggestion
Equivalent to \tcode{return x.\exposid{current_} <=> y.\exposid{current_};}.
```

File: source/ranges.tex [hunk 6]
```diff
@@ -4969,6 +5399,21 @@
 *(ranges::begin(E) + std::\linebreak{}min<D>(ranges::distance(E), F)))},
 except that \tcode{E} is evaluated only once.
 
+\item
+Otherwise, if \tcode{T} is
+a specialization of \tcode{ranges::repeat_view}\iref{range.repeat.view}:
+\begin{itemize}
+\item
+if \tcode{T} models \tcode{sized_range},
```
Comment (JohelEGP) [inline source/ranges.tex:5408]: Replace all `\tcode{sized_range}` with `\libconcept{sized_range}`

File: source/ranges.tex [hunk 7]
```diff
@@ -5400,6 +5845,22 @@
...
+then
+\begin{codeblock}
+views::repeat(*E.@\exposid{value_}@, ranges::distance(E) - std::min<D>(ranges::distance(E), F))
+\end{codeblock}
+except that \tcode{E} is evaluated only once;
+\item
+otherwise, ((void) F, decay-copy(E)),
```
Comment (JohelEGP) [inline source/ranges.tex:5861]:
```suggestion
otherwise, ((void) F, \placeholdernc{decay-copy}(E)),
```

**Comment by jensmaurer** [comment]
Note: This fails the automatic build, because [range.repeat.view] Apply modifications from P2494R2
refers to range.move.wrap, which is being renamed from range.copy.wrap by P2494R2, to be applied by #5710.

**Review by JohelEGP — CHANGES_REQUESTED** [review]

File: source/ranges.tex [hunk 8]
```diff
@@ -3202,6 +3210,429 @@
...
+\pnum
+\effects
+Equivalent to: \tcode{return to-unsigned-like(\exposid{bound_});}
+\end{itemdescr}
+
+\rSec3[range.repeat.iterator]{Class template \tcode{repeat_view::\exposid{iterator}}}
+
+\begin{codeblock}
+namespace std::ranges {
+  template<@\libconcept{copy_constructible}@ W, @\libconcept{semiregular}@ Bound = unreachable_sentinel_t>
```
Comment (JohelEGP) [inline source/ranges.tex:3379]:
```suggestion
  template<@\libconcept{move_constructible}@ W, @\libconcept{semiregular}@ Bound = unreachable_sentinel_t>
```

**Review by JohelEGP — CHANGES_REQUESTED** [review]

File: source/ranges.tex [hunk 9]
```diff
@@ -3202,6 +3210,428 @@
...
+    requires @\exposconcept{is-integer-like}@<Bound> || @\libconcept{same_as}@<Bound, unreachable_sentinel_t>
+  class repeat_view<W, Bound>::@\exposid{iterator}@ {
+  private:
+    using @\exposid{index-type}@ =                  // \expos
+      conditional_t<@\libconcept{same_as}@<Bound, unreachable_sentinel_t>, ptrdiff_t, Bound>;
+    const W * @\exposid{value_}@ = nullptr;         // \expos
```
[the remark this answers is not in the document]
Comment (JohelEGP) [reply source/ranges.tex:3385]: Do not miss https://github.com/cplusplus/draft/pull/5709#discussion_r939169605.

File: source/ranges.tex [hunk 10]
```diff
@@ -182,6 +182,14 @@
 
   namespace views { inline constexpr @\unspec@ iota = @\unspec@; }
 
+  // \ref{range.repeat}, repeat view
+  template<@\libconcept{copy_constructible}@ W, @\libconcept{semiregular}@ Bound = unreachable_sentinel_t>
+    requires (is_object_v<W> && @\libconcept{same_as}@<W, remove_cv_t<W>>
+      && (@\exposconcept{is-integer-like}@<Bound> || @\libconcept{same_as}@<Bound, unreachable_sentinel_t>))
```
[the remark this answers is not in the document]
Comment (jensmaurer) [reply source/ranges.tex:188]: fixed

File: source/ranges.tex (hunk 9 above)
[the remark this answers is not in the document]
Comment (jensmaurer) [reply source/ranges.tex:3385]: fixed

File: source/ranges.tex (hunk 4 above)
Comment (jensmaurer) [reply source/ranges.tex:3577]: No; as-is, this says that the "operator==" function call is equivalent to the following expression. Which seems reasonable.

File: source/ranges.tex (hunk 5 above)
Comment (jensmaurer) [reply source/ranges.tex:3588]: see above.

Also, if we add "return", we also need a colon before it.

File: source/ranges.tex (hunk 6 above)
Comment (jensmaurer) [reply source/ranges.tex:5408]: fixed

File: source/ranges.tex (hunk 4 above)
Comment (JohelEGP) [reply source/ranges.tex:3577]: I agree it is. But there's no precedent, AFAIK.

Comment (JohelEGP) [reply source/ranges.tex:3577]: And https://github.com/cplusplus/draft/wiki/Specification-Style-Guidelines says:
> For "Equivalent to", a plain expression should appear only if it is of type void;

File: source/ranges.tex (hunk 5 above)
Comment (JohelEGP) [reply source/ranges.tex:3588]: Added the colons.

File: source/ranges.tex [hunk 11]
```diff
@@ -3202,6 +3210,429 @@
...
+
+    @\exposid{movable-box}@<W> @\exposid{value_}@ = W();    // \expos, see \ref{range.move.wrap}
+    Bound @\exposid{bound_}@ = Bound();    // \expos
+
+  public:
+    repeat_view() requires @\libconcept{default_initializable}@<W> = default;
+
+    constexpr explicit repeat_view(const W& value, Bound bound = Bound())
+      requires @\libconcept{copy_constructible}@<W>;
+    constexpr explicit repeat_view(W && value, Bound bound = Bound());
```
Comment (JohelEGP) [inline source/ranges.tex:3259]:
```suggestion
    constexpr explicit repeat_view(W&& value, Bound bound = Bound());
```

File: source/ranges.tex [hunk 12]
```diff
@@ -3202,6 +3210,428 @@
...
+    // \ref{range.repeat.iterator}, class \tcode{range_view::\exposid{iterator}}
+    struct @\exposid{iterator}@;    // \expos
+
+    @\exposid{copyable-box}@<W> value_ = W();   // \expos, see \ref{range.copy.wrap}
+    Bound @\exposid{bound_}@ = Bound();    // \expos
+
+  public:
+    repeat_view() requires @\libconcept{default_initializable}@<W> = default;
+
+    constexpr explicit repeat_view(const W & value, Bound bound = Bound());
```
[the remark this answers is not in the document]
Comment (jensmaurer) [reply source/ranges.tex]: fixed

File: source/ranges.tex [hunk 13]
```diff
@@ -3202,6 +3210,429 @@
...
+\effects
+Equivalent to: \tcode{return \exposid{to-unsigned-like}(\exposid{bound_});}
+\end{itemdescr}
+
+\rSec3[range.repeat.iterator]{Class template \tcode{repeat_view::\exposid{iterator}}}
+
+\begin{codeblock}
+namespace std::ranges {
+  template<@\libconcept{copy_constructible}@ W, @\libconcept{semiregular}@ Bound = unreachable_sentinel_t>
+    requires @\exposid{is-integer-like}@<Bound> || @\libconcept{same_as}@<Bound, unreachable_sentinel_t>
```
Comment (JohelEGP) [inline source/ranges.tex:3380]: Mind https://github.com/cplusplus/draft/pull/5709#discussion_r939174850.

File: source/ranges.tex (hunk 8 above)
Comment (jensmaurer) [reply source/ranges.tex:3379]: fixed

**Review by JohelEGP — CHANGES_REQUESTED** [review]

File: source/ranges.tex [hunk 14]
```diff
@@ -3202,6 +3210,428 @@
...
+\effects
+Equivalent to: \tcode{return to-unsigned-like(\exposid{bound_});}
+\end{itemdescr}
+
+\rSec3[range.repeat.iterator]{Class template \tcode{repeat_view::\exposid{iterator}}}
+
+\begin{codeblock}
+namespace std::ranges {
+  template<@\libconcept{copy_constructible}@ W, @\libconcept{semiregular}@ Bound = unreachable_sentinel_t>
+    requires @\exposconcept{is-integer-like}@<Bound> || @\libconcept{same_as}@<Bound, unreachable_sentinel_t>
```
[the remark this answers is not in the document]
Comment (jensmaurer) [reply source/ranges.tex]: fixed

File: source/ranges.tex [hunk 15]
```diff
@@ -3202,6 +3210,430 @@
...
+    const W* @\exposid{value_}@ = nullptr;          // \expos
+    @\exposid{index-type}@ @\exposid{current_}@ = @\exposid{index-type}@(); // \expos
+
+    constexpr explicit @\exposid{iterator}@(const W* value, @\exposid{index-type}@ b = @\exposid{index-type}@());   // \expos
+
+  public:
+    using iterator_concept = random_access_iterator_tag;
+    using iterator_category = random_access_iterator_tag;
+    using value_type = W;
+    using difference_type = conditional_t<@\exposid{is-signed-like}@<@\exposid{index-type}@>,
```
Comment (hewillk) [inline source/ranges.tex:3395]: What is _`is-signed-like`_? Should this be _`is-signed-integer-like`_? I suspect this is a typo in the paper.

File: source/ranges.tex [hunk 16]
```diff
@@ -3202,6 +3210,430 @@
...
+
+\begin{itemdescr}
+\pnum
+\expects
+If \tcode{Bound} is not \tcode{unreachable_sentinel_t}, $\tcode{b} \ge 0$.
+
+\pnum
+\effects
+Initializes \exposid{value_} with \tcode{value} and
+\exposid{bound_} with \tcode{b}.
```
Comment (hewillk) [inline source/ranges.tex:3438]: _`iterator`_ does not have a member _`bound_`_, should this be _`current_`_? I suspect this is a typo in the paper.

File: source/ranges.tex (hunk 13 above)
Comment (tkoeppe) [reply source/ranges.tex]: @JohelEGP Could you say what exactly needs changing here?

File: source/ranges.tex (hunk 16 above)
Comment (tkoeppe) [reply source/ranges.tex:3562]: @jwakely, @CaseyCarter: or would you like to fix this immediately?

File: source/ranges.tex (hunk 15 above)
Comment (tkoeppe) [reply source/ranges.tex:3519]: Yes, I think so, thanks!

File: source/ranges.tex (hunk 13 above)
Comment (JohelEGP) [reply source/ranges.tex]: The fix has been applied. This was meant to call attention to the linked comment which had been outdated by an unrelated change.

Comment (tkoeppe) [reply source/ranges.tex]: Aha, excellent -- then I think we're ready here!

**Comment by JohelEGP** [comment]
The commit linked by https://github.com/cplusplus/draft/pull/5709#discussion_r939200940 was related to https://github.com/cplusplus/draft/pull/5710, right?

From https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2022/p2494r2.html:
> Additionally, this paper proposes to modify the changes in P2474 in this same way. If both this paper and P2474 are accepted, we should make views::repeat have the same constraints as views::single.

Then it has [5. Modifications to P2474](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2022/p2494r2.html#wording-p2474).

It seems these were reverted and need to be redone.

**Comment by tkoeppe** [comment]
Yes yes: it's not reverted, just not on this branch anymore. I still have the commit :-)

File: source/ranges.tex (hunk 16 above)
Comment (tkoeppe) [reply source/ranges.tex]: Actually, since this seems like a simple error, let's just fix it right now.

**Pull request merged**